src/npip.js

```diff
@@ -5,27 +5,71 @@
 const fs = require("fs");
 const path = require("path");
 const { findPackage, spawnPython } = require("./api");
-const { promisify } = require("./promisify");
 
-const stat = promisify(fs.stat);
-
-const installPip = (pkg) => {
-  return stat(path.join(pkg.dir, "python_modules")).catch(error => {
+const installPip = async (pkg) => {
+  try {
+    await fs.promises.stat(path.join(pkg.dir, "python_modules"));
+    return;
+  } catch (error) {
     if (error.code !== "ENOENT")
       throw error;
-    console.log("No python_modules directory; installing pip locally if needed.");
-    return spawnPython([path.join(__dirname, "get-pip.py"), "--user", "--quiet"], {
-      package: pkg,
-      interop: "status",
-      spawn: {
-        stdio: "inherit",
-      }
-    }).then(code => {
-      console.log("Successfully completed pip check.");
-      return code;
+  }
+
+  console.log("No python_modules directory; installing pip locally if needed.");
+  const opt = {
+    package: pkg,
+    interop: "status",
+    spawn: {
+      stdio: "inherit"
+    }
+  };
+
+  // Check if pip is already installed.
+  try {
+    await spawnPython(["-m", "pip", "-V"], opt);
+    return;
+  } catch (e) { }
+
+  // Try to install via ensurepip.
+  try {
+    await spawnPython(["-m", "ensurepip", "--upgrade", "--user"], opt);
+    console.log("Successfully completed pip check.");
+    return;
+  } catch (e) { }
+
+  // run get-pip.py.
+  try {
+    await spawnPython([path.join(__dirname, "get-pip.py"), "--user", "--quiet"], opt);
+    console.log("Successfully completed pip check.");
+    return;
+  } catch (error) {
+    console.error("Failed to install pip locally.\n" + JSON.stringify(error));
+  }
+
+  // Last resort: download pip and run it.
+  const tmpFile = path.join(pkg.dir, "get-pip.py");
```

Comment:
I think get-pip.py should be downloaded to a temporary directory rather than the package directory. Not _the_ temporary directory, because of potential [security issues](https://en.wikipedia.org/wiki/Symlink_race). I think this would do:
```js
await mkdtemp(path.join(os.tmpdir(), 'npip-'));
```
And delete the directory when finished.